openedx/core/djangoapps/content_staging/models.py

```diff
@@ -0,0 +1,125 @@
+"""
+Models for content staging (and clipboard)
+"""
+import logging
+
+from django.contrib.auth import get_user_model
+from django.core.exceptions import ValidationError
+from django.db import models
+from django.utils.translation import gettext_lazy as _
+from opaque_keys.edx.django.models import UsageKeyField
+from opaque_keys.edx.keys import LearningContextKey
+
+from openedx.core.djangoapps.content.course_overviews.api import get_course_overview_or_none
+
+log = logging.getLogger(__name__)
+
+User = get_user_model()
+
+
+class StagedContent(models.Model):
+    """
+    Each StagedContent instance represents a "piece" of content (e.g. a single
+    XBlock, or a single Unit, or a single Subsection of a course) that is not
+    currently part of any course or library, but which is available to be copied
+    into a course or library.
+
+    Use as a clipboard: for any given user, the most recent row with
+    purpose=CLIPBOARD is the "current" clipboard content. But it can only be
+    pasted if its status is READY.
+    """
+
+    class Meta:
+        verbose_name_plural = _("Staged Content")
+
+    class Purpose(models.TextChoices):
+        """ The purpose of this staged content. """
+        CLIPBOARD = "clipboard", _("Clipboard")
```

Comment:
Does StagedContent really need this? StagedContent is otherwise a lower-level model that UserClipboard makes use of. It seems odd for it to have this kind of knowledge.

Reply:
I know, but otherwise it's hard to tell what the StagedContent was for. `UserClipboard` has a OneToOne relationship to User, which means that when a user copies something new, there is no "old" UserClipboard entry. So when a user copies something then copies something else, you end up with"stranded" `StagedContent` entries that aren't pointed to by `UserClipboard` and so without this field, there is no information about their purpose. It's impossible to tell from the database if they were used for clipboard or for some other thing (partial import, template, ... ?)

Now, if I make it so that UserClipboard always immediately deletes its old StagedContent rows, that would solve the problem, but that makes copying slower unless it's done asynchronously. And if it's done asynchronously, it might fail and then you'd get "stranded" StagedContent rows with no information about what they were used for. At least, that was my thought process. Let me know if you think that's not important.

Another option is to leave the purpose field but don't make it an enum, so that it can be flexibly used without knowing all the possible apps that will use it. Or I can just delete it any rely on the fact that for now all stagedcontent is clipboard content?

Reply:
> Another option is to leave the purpose field but don't make it an enum, so that it can be flexibly used without knowing all the possible apps that will use it. Or I can just delete it any rely on the fact that for now all stagedcontent is clipboard content?

I do think we should kick off the async deletion of the old StagedContent as soon as a new thing has started copying, but I take the point that deletion could always fail. I have a mild preference to just leave it a freeform field then (ditching the Enum as you suggest here). But I don't feel strongly about it.